File: Java/binary_tree.java
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

/**
 * Simply, a binary tree.
 * Author: Michael Cowie
 */
public class Main {

   public static void main(String[] args){
      BinaryTree binaryTree = new BinaryTree();
      binaryTree.add(5);
      binaryTree.add(3);
      binaryTree.add(2);
      binaryTree.add(10);
      for(Integer nodeVal : binaryTree.bfsTraversal()){
         System.out.print(nodeVal + " "); //5 3 10 2
      }
   }

}

class Node {
   Integer value;
   Node left = null;
   Node right = null;
   public Node(int value){
      this.value = value;
   }
}

class BinaryTree {
   Node root;

   public BinaryTree(){
      this.root = null;
   }

   public void add(int value){
      if(this.root == null){
         this.root = new Node(value);
      } else{
         add_helper(this.root, value);
      }
   }

   public void add_helper(Node node, int value){
      //Check if traversing left
      if(value < node.value){
         //Check if leaf node
         if(node.left == null){
            node.left = new Node(value);
         } else{
            add_helper(node.left, value);
         }
         //Check if traversing right
      } else if(value > node.value){
         //Check if leaf node
         if(node.right == null){
            node.right = new Node(value);
         } else{
            add_helper(node.right, value);
         }
      }
   }

   public List<Integer> bfsTraversal(){
      List<Integer> traversalValue = new ArrayList<>();
      ArrayList<Node> queue = new ArrayList<>(Arrays.asList(this.root));
      while (queue.size() > 0){
         Node nodeToPop = queue.remove(0);
         if (nodeToPop != null) {
            traversalValue.add(nodeToPop.value);
            queue.add(nodeToPop.left);
            queue.add(nodeToPop.right);
         }
      }

      return traversalValue;
   }
}

File: Python/Algorithms/binary_tree.py
class Node():
    def __init__(self, value):
        self.value = value
        self.left = None
        self.right = None
        
#My binary tree implementation does not allow duplicates        
class Tree():
    def __init__(self):
        self.root = None
    
    def add(self, value):
        if self.root is None:
            self.root = Node(value)
        else:
            self.add_helper(value, self.root)
            
    #Recursive method reaching all nodes     
    def add_helper(self, value, node):
        #Check left node
        if value < node.value:
            #Check if it is a leaf node
            if node.left is None:
                node.left = Node(value)
            else:
                self.add_helper(value, node.left)
        #Check right node        
        elif value > node.value:
            if node.right is None:
                node.right = Node(value)
            else:
                self.add_helper(self, value, node.right)
                
    def print_tree(self):
        if self.root != None:
            bfs = self.bfs_traversal(self.root)
            print(bfs)
            #incomplete print of tree
            
    def bfs_traversal(self, node):
        traversal_order = []
        queue = [node]
        while len(queue) != 0:
            node_popped = queue.pop(0)
            if node_popped != None:
                traversal_order.append(node_popped.value)
                queue.append(node_popped.left)
                queue.append(node_popped.right)
        return traversal_order

                   
tree = Tree()
tree.add(5)
tree.add(3)
tree.add(1)
tree.add(7)
tree.add(4)
tree.print_tree()
